### ISSUE TITLE
system for tags/+releases; namespaces and publish schedules

### ISSUE THREAD

**Issue by meowsbits** [multi-geth/multi-geth#80]
Rel #35 

There are now [(as proposed)](https://github.com/multi-geth/multi-geth/pull/79) two parallel branches

1. master
2. release/X.Y

Upstream, releases are all tagged on the `release/X.Y` branch, while that branch is significantly diverged from `master`. 

![Screenshot_20190517_111704](https://user-images.githubusercontent.com/45600330/57941766-67e48300-7895-11e9-94d3-0372cc2f6d33.png)

It seems to me that current master is typically quite stable (at least relative to it's releasing cousin), and at this point has features that want releasing (again, eg. `rpc.discover`).

Backporting features like this from master to release/X.Y is [extra and "backwards walking" work, and exposes extra complexity.](https://github.com/multi-geth/multi-geth/pull/79#issue-279842571) This is not handled very elegantly upstream, either, and I'm not convinced that a tag on a `release/X.Y` branch having the name `-stable-` in it really means it is actually (more) stable vs. tagging on master.

Questions:

- On which branch or branches should we create tags?
  + How should we handle these relative to upstream?
- How should we name tags (respective of their relative branches)?
- When should we tag tags?
- When should we publish releases?

Considerations:

- Creating differentiated tags across divergent branches `master` and `release-` creates extra complexity, both for developer and end-user. I'm not sure this extra complexity is at this point warranted.
  + On the contrary, given two 'canonical' branch paths, there is simplicity to an approach that acknowledges those parallel-ish paths, and instead of attempting to prioritize or reconcile them, simply uses a "two-tag" system. 
- [Lately there were some notable bugs and errors in the upstream published tag/releases](https://github.com/ethereum/go-ethereum/releases). This is a signal that "stability" may be an exaggeration.
  + This should inform our process and patterns for echoing upstream tag/release publishing.

__Prop 1 "Two-tag"__: Use `v1.2.3-$branch_name` tag names, where `$branch_name` will be either `master` or `release`. Semver values would be incremented _independently_ of each other. 
  + Release tags (and releases) would echo upstream releases (and schedules) with, say, a 1-2 week lag time to wait for bugs to get caught. The major concept is to publish a version in near-lockstep with upstream published version (w/r/t development/features and timing). 
  + Master tags (and releases) would work to signal notable changes, fixes, and upgrades, that may or may not echo upstream developments.

Using a "third-arm beta" branch/tag system based off of master (like upstream theoretically does with `release/X.Y`), but behaving independently of upstream) would be a way to manage publishing "stabler" versions, much as @soc1c [proposed](https://github.com/multi-geth/multi-geth/issues/35#issuecomment-465021376). Although of course this poses a potential situation where instead of demystifying this scenario, it could snowball the complexity.

Thoughts?

**Comment by tzdybal** [comment]
In my opinion, we should:
- limit to following single upstream branch - `master`
- release independently from upstream (both release numbering and dates)
- release as often as needed (after each feature, fix, new chain, etc)

Rationale:
- we have no control over stability of "stable" versions from upstream
- changes between release branches and master are significant, which makes backporting slow and painful process
- we should write actual code, and not spend most of the time mastering git merge and rebase process

**Comment by sorpaas** [comment]
One of the reason that upstream's using `release/1.x` is because it does not want any small breakage of backward compatibility on JSON-RPC interface or command line API, so only fixes and important PRs are backported. I don't think we need such strong guarantees for multi-geth at this moment.

**Comment by meowsbits** [comment]
I'm in agreement. 

I'm going to merge #79  since it seems like there is value in having a comparable multi-geth-ized 1.8 branch, since it will probably be the near the last of the 1.8 sequence, and would just be wasted code otherwise. This doesn't mean we'd have to tag anything there necessarily, just that it would be there in case it ever becomes relevant (can't hurt, could help).

I'll propose then that we push our first "independent" tag with a minor-level bump to 1.9.0 at master's d1975dfe9a1420de0335d246b6c5cc34e0e8cf46, given that our 'new functionalities' include _at least_ `rpc.discover` which is of notable value for developers. 


> Given a version number MAJOR.MINOR.PATCH, increment the:
> 
>     MAJOR version when you make incompatible API changes,
>     MINOR version when you add functionality in a backwards-compatible manner, and
>     PATCH version when you make backwards-compatible bug fixes.

**Comment by meowsbits** [comment]
And w/r/t to #81 - feat: Musicoin, I'd suggest we include that with 1.10 or similar, but push priority on review of that down a little to next release.

**Comment by meowsbits** [comment]
Because version values are hardcoded, we'll need a pattern for bumping them.

https://github.com/multi-geth/multi-geth/blob/master/params/version.go#L23-L29

How about "mini-branches", eg. `release/1.9.0`, that are only 1 commit beyond master, which should only include those version adjustments. Respective tags are then made at those tiny branch tips.

**Comment by tzdybal** [comment]
I agree with "mini-branches" concept. Clean and simple. And :+1: w/r/t numbering.
